capa/main.py

```diff
@@ -303,29 +303,43 @@ class UnsupportedRuntimeError(RuntimeError):
     pass
 
 
-def get_extractor_py3(path, format, disable_progress=False):
-    from smda.SmdaConfig import SmdaConfig
-    from smda.Disassembler import Disassembler
+def get_extractor_py3(path, format, backend, disable_progress=False):
+    if backend == "smda":
+        from smda.SmdaConfig import SmdaConfig
+        from smda.Disassembler import Disassembler
 
-    import capa.features.extractors.smda
+        import capa.features.extractors.smda
 
-    smda_report = None
-    with halo.Halo(text="analyzing program", spinner="simpleDots", stream=sys.stderr, enabled=not disable_progress):
-        config = SmdaConfig()
-        config.STORE_BUFFER = True
-        smda_disasm = Disassembler(config)
-        smda_report = smda_disasm.disassembleFile(path)
+        smda_report = None
+        with halo.Halo(text="analyzing program", spinner="simpleDots", stream=sys.stderr, enabled=not disable_progress):
+            config = SmdaConfig()
+            config.STORE_BUFFER = True
+            smda_disasm = Disassembler(config)
+            smda_report = smda_disasm.disassembleFile(path)
+
+        return capa.features.extractors.smda.SmdaFeatureExtractor(smda_report, path)
+    else:
+        import capa.features.extractors.viv
+
+        with halo.Halo(text="analyzing program", spinner="simpleDots", stream=sys.stderr, enabled=not disable_progress):
+            vw = get_workspace(path, format, should_save=False)
 
-    return capa.features.extractors.smda.SmdaFeatureExtractor(smda_report, path)
+            try:
+                vw.saveWorkspace()
+            except IOError:
+                # see #168 for discussion around how to handle non-writable directories
+                logger.info("source directory is not writable, won't save intermediate workspace")
 
+        return capa.features.extractors.viv.VivisectFeatureExtractor(vw, path)
 
-def get_extractor(path, format, disable_progress=False):
+
+def get_extractor(path, format, backend="vivisect", disable_progress=False):
```

Comment:
lets make `backend` a required parameter.

i think we should also use constants, rather than string literals.

Reply:
@williballenthin 


> lets make backend a required parameter.

doesn't that make it more difficult for the user? I mean, what if you just want to get the capabilities

> i think we should also use constants, rather than string literals.

I am not sure what you mean with constants. 😕 Can you please elaborate?

Reply:
> doesn't that make it more difficult for the user? If I use capa to get a program's capabilities, I don't really care which backend is used. I would prefer to get a backend recommendation/default (preferably the best one).

right, lets provide a default backend when parsing the cli arguments, but i don't see a good reason for the backend not be to a required argument to this internal routine.

Reply:
> I am not sure what you mean with constants. 😕 Can you please elaborate?

```py
BACKEND_VIV = "vivisect"
BACKEND_SMDA = "smda"
```

Reply:
@williballenthin 
> but i don't see a good reason for the backend not be to a required argument to this internal routine.

ah! that's what you meant! I gave it a default because we are calling this function in 4 scripts. Should I require it here and provide a backend in the scripst?

Reply:
yes, please do that

Reply:
Note that this function is also called from `capa/features/freeze.py`. I think it make sense to provide a CLI option there as well.